### PR TITLE
Weaken the superclass on `Commutative`

### PR DESCRIPTION
**Pull request by Ericson2314**
That it was something as strong as `Fractional` was a relic of the `Group` superclass to `Abelian` I stripped away to make this package.

However, `Num`'s suggested laws do not come with commutativity for
`(*)`. Should we do this anyways? or should we have a
```haskell
class Num a => CommutativeMultiplication a
```
helper class so we can do
```haskell
instance CommutativeMultiplication a => Commutative (Product a)
```
?

CC @cgibbard @madeline-os @endgame @Taneb

@Taneb in particular, note that the safer `CommutativeMultiplication` would be a breaking change for `groups`.

### CODE REVIEW

**Comment by endgame** [comment]
Hm. Do people give matrices a `Num` instance where `(*)` is used as matrix multiplication? What other common noncommutative multiplications do people use?

**Comment by Ericson2314** [comment]
I'm not sure, maybe the CLC should amend the laws to also require commutivity.

**Comment by endgame** [comment]
I have an example of noncommutative `(*)` in a "real" lib: https://hackage.haskell.org/package/linear-1.21.8/docs/src/Linear.Quaternion.html#line-243

**Comment by endgame** [comment]
An alternative: what if we defined `newtype CommutativeProduct a = CommutativeProduct { getCommutativeProduct :: a }`, with `Semigroup`, `Monoid`, `Commutative`, and `Group` instances, but **not** `instance Num a => Commutative (CommutativeProduct a)`? We'd have to enumerate a lot of primitive types whose products are commutative, but we don't claim an instance for e.g., quaternions by mistake.

**Comment by Taneb** [comment]
In my mind, all the numeric instances for `Semigroup`/`Monoid`/`Group` are to be taken with a strong pinch of salt in any case. Floating point addition isn't associative, for example. My preferred end-result would be without a `CommutativeProduct` and with a `Num` constraint on `Commutative (Product a)` personally. However, if other people decide that a `CommutativeProduct` class is the way to go, I have no objections.

**Comment by Ericson2314** [comment]
@Taneb Well, what do you think about that `Quaternion` example linked above? That is a much more *intensional* non-commutativity than rounding errors messing up associativity.

**Comment by endgame** [comment]
An unserious suggestion to shake up the discussion: what if we didn't provide instances for `Product a` at all? If you want to commute products, take `exp`, use `Sum` and then take `log`?

Maybe someone else can come up with a better off-beat answer?

**Comment by Ericson2314** [comment]
Ah I might not have mentioned it, but see https://github.com/blamario/monoid-subclasses/pull/35/files#diff-bd6dc0a969919d08c6410c5310ee389410e359c2e188245eb1b5a5890d31a6cbL201 so that library expects *something* along these lines.

**Comment by endgame** [comment]
So here's another thorn: if we want to require a specific subclass which implies commutative `(*)`, we have multiple candidates:

* `Floating` customarily expects `(+)`, `(*)`, and `exp` to define an exponential field, and by definition an exponential field is a commutative ring with additional conditions;
* `Integral` customarily expects its instance to define a Euclidean domain, and Euclidean domains are also commutative rings with additional conditions.
* `Fractional` does not: it is expected to define a division ring, which is commutative iff it is also a field.

If we did provide `instance Num a => Commutative (Product a)` we could ask `linear` to add an overlapping instance with type error:

```
instance
  {-# OVERLAPS #-}
  (Semigroup a, TypeError ('Text "Multiplication is not commutative on Quaternions")) =>
  Commutative (Product (Quaternion a))
```

But this means we break anyone who defines a noncommutative `(*)`, and we have to find them, beg forgiveness, and ask them to depend on us. This isn't just `Quaternion` - anyone who writes a `Num` instance for square matricies would also be affected, and anyway we can't provide the overlapping instance for `Quaternion` because `linear` depends on `lens` and by extension half the known universe. So we have to ask more people to depend on this library.

At the moment, I'm in favour of `class Num a => CommutativeProduct a` and `instance (CommutativeProduct a, Semigroup a) => CommutativeSemigroup (Product a)`, unless someone finds something better.

**Comment by endgame** [comment]
Another possible point in favour of a second `CommutativeMultiplication` class: People don't define new `Num` classes _that_ often in my experience. We can write out instances for everything in `base`, which will probably be a large subset of the commutative `Num` instances that anyone will ever care about.

**Comment by Ericson2314** [comment]
> Another possible point in favour of a second `CommutativeMultiplication` class: People don't define new `Num` classes _that_ often in my experience. We can write out instances for everything in `base`, which will probably be a large subset of the commutative `Num` instances that anyone will ever care about.

Yeah this what I was thinking too.

**Comment by endgame** [comment]
It's been two weeks. Are we happy to commit to a `CommutativeMultiplication` class?

**Comment by madeline-os** [comment]
I am canvassing community members for their opinion. I think the extra class makes sense, but I want more perspectives on the issue.

File: ChangeLog.md [hunk 1]
```diff
@@ -2,8 +2,9 @@
 
 ## Unreleased
 
-- `Commutative (Product a)` now merely requires
-  `Num a`.
+- `Commutative (Product a)` now merely requires `CommutativeProduct a`.
```
Comment (endgame) [inline ChangeLog.md]:
```suggestion
- `Commutative (Product a)` now requires `CommutativeProduct a`.
```
Perhaps also mention some things we found in our discussion?

Comment (Ericson2314) [reply ChangeLog.md]: Yeah I just did the bare minimum so the change log wouldn't be wrong. Feel free to propose what it ought to say from scratch if you like.

Comment (Ericson2314) [reply ChangeLog.md]: Marked as draft because there are many more Num instances in base to add.

Comment (endgame) [reply ChangeLog.md]: #5 fleshes out the changelog entry, and is set to merge into this branch.

**Comment by Ericson2314** [comment]
Nicely worded! Perhaps some of the change log entry should become haddocs for the class, instead or in addition?

**Comment by endgame** [comment]
Good idea, you should copy and massage some of the changelog entry into a haddock.

**Comment by Ericson2314** [comment]
@blamario and @Taneb if you could review this, that would be much appreciated.

**Review by blamario — APPROVED** [review]
I generally prefer class names to be adjectives, and in this case doubly so because the name looks like a variant of `Product` which is a data type. Unfortunately I can't offer anything better than `CommutativelyMultiplicative` which is quite a mouthful so I'll approve the PR as it stands.

**Review by Taneb — APPROVED** [review]
I'm happy with this design

**Comment by Ericson2314** [comment]
Thank you both!

**Pull request merged**